src/tools/ilasm/ILAssembler/PreprocessedTokenSource.cs

```diff
@@ -0,0 +1,278 @@
+                    }
+                    if (nextToken.Type == CILLexer.PP_ENDIF)
+                    {
+                        numNestedPreprocessorBlocks--;
+                    }
+                    continue;
+                }
+
+                if (nextToken.Type == CILLexer.PP_ELSE)
+                {
+                    if (elseCase)
+                    {
+                        ReportPreprocessorSyntaxError(nextToken);
+                        return NextTokenWithoutNestedEof(errorOnEof: true);
+                    }
+
+                    _activeIfDefBlocks.Push((var, Defined: !expectedDefined, IsElse: true));
+                    var (source, activeIfDef) = _includeSourceStack.Pop();
+                    _includeSourceStack.Push((source, ++activeIfDef));
+
+                    return NextTokenWithoutNestedEof(errorOnEof: true);
+                }
+                if (nextToken.Type == CILLexer.PP_IFDEF || nextToken.Type == CILLexer.PP_ENDIF)
+                {
+                    numNestedPreprocessorBlocks++;
+                }
+            }
+
+            // If we're skipping an inactive else case, then we're still tracking the active ifdef block.
+            if (elseCase)
+            {
+                var (source, activeIfDef) = _includeSourceStack.Pop();
+                _includeSourceStack.Push((source, --activeIfDef));
+            }
+            return NextTokenWithoutNestedEof(ActiveIfDefBlocksInCurrentSource != 0);
+        }
+
+        public event Action<string, int, int, string>? OnPreprocessorSyntaxError;
+
+        private void ReportPreprocessorSyntaxError(IToken token)
+        {
+            string text = token.TokenSource.InputStream.GetText(Interval.Of(token.StartIndex, token.TokenSource.InputStream.Index));
+            string msg = "preprocessor syntax error at: '" + GetErrorDisplay(text) + "'";
+            OnPreprocessorSyntaxError?.Invoke(token.TokenSource.SourceName, token.StartIndex, token.StopIndex - token.StartIndex, msg);
+        }
+
+        private static string GetErrorDisplay(string s)
+        {
+            StringBuilder stringBuilder = new StringBuilder();
+            var array = s.AsSpan();
+            foreach (char c in array)
+            {
+                stringBuilder.Append(GetErrorDisplay(c));
+            }
+
+            return stringBuilder.ToString();
+        }
+
+        private static string GetErrorDisplay(int c)
+        {
+            string result = ((char)c).ToString();
+            switch (c)
+            {
+                case -1:
+                    result = "<EOF>";
+                    break;
+                case 10:
+                    result = "\\n";
+                    break;
+                case 9:
+                    result = "\\t";
+                    break;
+                case 13:
+                    result = "\\r";
+                    break;
+            }
+
+            return result;
+        }
```

Comment:
Nit: it can use short syntax.

```suggestion
        private static string GetErrorDisplay(int c) => c switch
        {
            -1 => "<EOF>",
            10 => "\\n",
            9  => "\\t",
            13 => "\\r",
            _  => ((char)c).ToString()
        };
```